### DISCUSSION TITLE
TI API - New potential requirement: Device Source IP Port

### DISCUSSION THREAD

**Discussion by FabrizioMoggio** [camaraproject/EdgeCloud#230] [Ideas]
it is possible for a Device to setup more data flows using specific Source IP Ports. Currently the TI API doesn't support a Device with source port number.

Are there use cases requiring to support this feature?

**Comment by javierlozallu** [comment]
I will add this topic to the agenda to ask for feedback, @FabrizioMoggio if you can explain there a bit what this new requirement is about it would be perfect.

**Comment by babunkj** [comment]
There are scenarios where a device has multiple traffic/streams which are identified based on Port numbers. An example is the camera with low and high resolution traffic which have different port numbers allocated to each. There are use cases where only one or selected streams from the device need to be routed to a specific Edge or Application Server. This identification of the traffic can be done by introducing source port numbers in the request, wherein which the Application can specify which stream (corresponding to the ports mentioned) need to be routed / influenced.

**Reply by gunjald** [reply]
I think in current version the device is represented by its IPV4 address and the array of port and protocol associated it with which i will assume as the port of the edge application server based on my interpretation of the trafficFilters description. 
In this case the appId in the API points to a single application while the ports belongs to EAS (appId) and not to device as source port.

So even though the traffic can be filtered but i will assume that the filtered traffic will be routed to the edge application identified by appId but the API doesnt support routing a specific traffic flow to different applications or EAS in my opinion.

**Reply by FabrizioMoggio** [reply]
@gunjald that is correct. Actually Port Number is only for the Destination (the EAS) and not for the Source (the Device). This is way I was proposing this new requirement to add the support for Device Source IP Port. A possible use case is actually provided by you in this discussion:

https://github.com/camaraproject/EdgeCloud/issues/123#issuecomment-2114633531

Am I wrong?

**Reply by gunjald** [reply]
I think for the sake of clarity we need to define the terms The Device, Destination, Source, Port etc. It becomes difficult to model which terms mean what at times. For example when we say Source IP does it mean device IP? When we say port where do we relate it to i.e. to Device or to EAS?

Also, my suggestion would be to refer to source IP, source Port, and destination IP , destination port just to ensure they are being taken up together and can form a unique key.

**Comment by FabrizioMoggio** [comment]
Currently we have in the TrafficInfluence resource some attibutes related to the Destination EAS and some others related to the Source that is the smartphone:

-Device: here we identify the Source
-trafficFilters: here we have the Destination port and protocol

With the goal to manage a Source port, we can modify Device adding  a new "devicePort" parameter, maybe in Ipv4Address and Ipv6Address definition.

or even better we can define two trafficFilters:

destintionTrafficFilters and sourceTrafficFilters. So we don't need to modifiy Device or Ipv4Address and Ipv6Address.

I propose to introduce:

destintionTrafficFilters and sourceTrafficFilters

**Reply by FabrizioMoggio** [reply]
I'm wondering:
- destintionTrafficFilters: protocol and port toward the EAS
- sourceTrafficFilters: only port? protocol must be the same set in destintionTrafficFilters?

**Reply by gunjald** [reply]
I think introduction of destintionTrafficFilters and sourceTrafficFilters will help as it clearly segregate the source and destination entities. But we will also need to explain the relationship between the set of IP addresses used in device->ipv4Address/ipv6Address parameters and the ones to be used in sourceTrafficFilters if we define so. And the possible addresses, their nature and how the application developers can obtain it specially the private addresses if they are also valid for the API.

I think we need more discussion within the group as suggested by @babunkj

**Reply by FabrizioMoggio** [reply]
@gunjald do you suggest to discuss this in the Edge Cloud regular meeting or to set up a TI API specific call? Is this topic of interest for the whole group?

**Reply by FabrizioMoggio** [reply]
One more comment, for the TI API, the invoker is usually the App Backend that is aware of the public IP for the Device so, for this API we shouldn't have any problem. Anyway this must be well described in the documentation. I don't see the TI API ever using private IP and ports for the Device, only the public ones are used.

**Reply by gunjald** [reply]
One point on IP, port mapping for specific flows in my opinion is that determining the mapping between internal IP:Port with the NATed IP:Port may be a bit of work as the mapping at NAT gateway will be performed dynamically. So if we can provide some references to help promote the understanding of parameters nature and how reliably they can be determined will really help speedy development.

**Comment by babunkj** [comment]
I agree, introducing destintionTrafficFilters and sourceTrafficFilters will bring more clarity on how the application can define 'what to influence'. They can be such that application can create them with minimum information that it may have.

**Comment by FabrizioMoggio** [comment]
About the improvement in the Documentation that is needed, we can start looking at:

Commonalities have this document on UE identification: https://github.com/camaraproject/Commonalities/blob/main/documentation/UE-Identification.md#af-specific-ue-external-identifier

**Reply by FabrizioMoggio** [reply]
The UE Identification document from Commonalities is too technical, I suggest to try to provide some simple suggestion and explanation in the TI API Documentation

**Reply by babunkj** [reply]
Yes, the UE Identification document in Commonalities looks more of a reference document. 
We can consider the data types given in https://github.com/camaraproject/Commonalities/blob/main/artifacts/CAMARA_common.yaml for the device Identifier types. The explanation of device Identifiers in the yaml can be a simple one.

**Reply by FabrizioMoggio** [reply]
@babunkj very good suggestion. I will update the YAML according to CAMARA_common.yaml and provide some further, simple, hint in the documentation section of the YAML.

**Reply by FabrizioMoggio** [reply]
in the CAMARA_common.yaml, Device can be identified by:

   properties:
    publicAddress:
     $ref: "#/components/schemas/SingleIpv4Addr"
    privateAddress:
     $ref: "#/components/schemas/SingleIpv4Addr"
    publicPort:
     $ref: "#/components/schemas/Port"

So it already has the Port. Adopting this we can avoid introducing sourceTrafficFilters.

Furthermore that definition includes both Private and Public Address specifying that the public one must always be specified. So I think that this covers also our needs.

So I propose to Update Device according to CAMARA_common.yaml. This provides explanation on Public and Private IPs. It includes source Port covering this New Requirement we are discussing here. We don't need to modify trafficFilters that is still used for the EAS.

**Reply by FabrizioMoggio** [reply]
This solution should cover Issue: https://github.com/camaraproject/EdgeCloud/issues/123

**Reply by babunkj** [reply]
The publicAddress and publicPort attributes together or the privateAddress alone will be sufficient to identify a device. The device port is additional info used to identify the specific traffic from the device. Thus, my suggestion is to treat the device identifier and traffic filters separately. It would be appropriate to put the device port info as part of traffic filter. Moreover, there are cases when the application wants to influence a different traffic from the device than the one it is connected to. This is the situation, for example, when an application wants to monitor a low resolution traffic and based on its analysis, influence a high resolution traffic. 

In a few other situations, if the application invoker is a trusted one and inside the NAT, it will be providing only privateAddress as the device identifier. In such cases, the application needs to specify the device port in TrafficFilter in order to identify the traffic to be influenced. 

So, I would suggest having the destintionTrafficFilters and sourceTrafficFilters where the device port can be provided in the sourceTrafficFilters.

**Reply by FabrizioMoggio** [reply]
@babunkj you are right. That was not clear to me until now: 

- public IP port: to identify the device under NAT
- protocol port: to identify a specific application flow on that device

So: we adopt the new Device definition from CAMARA_common.yaml, we introduce sourceTrafficFilters and we rename the old one into destinationTrafficFilters

Please note the on Device there is still an ongoing discussion:  https://github.com/camaraproject/Commonalities/issues/171

**Reply by gunjald** [reply]
I have a question to ask on "public IP port: to identify the device under NAT". As per my understanding a device will have a private IP address assigned by mobile core to it. But the port in "Public IP:Port" is something which will be specific to a traffic flow going out from the mobile network to external applications and that too will be associated to a specific protocol e.g. TCP, UDP. 

So how do we explain the meaning of port in "public IP port": to application developer as it is a dynamic port which can keep changing as the NAT gateway performs the mapping of internal port with private IP dynamically with a possibly different public port at different time and based on events like idle timeouts, location change resulting into dataplane change etc.

From application provider perspective he should be given network agnostic view in terms of what all scenario he should observe and what all action the application may need to take on those events e.g. determining IP:Port mapping changed and earlier mapping becomes invalid. So how developer can detect such events and initiate the API calls again.

But these are just my thoughts and would like to understand if the given concepts or parameter usage needs to viewed differently?

**Reply by FabrizioMoggio** [reply]
I'm wondering if it is not better, for the Developer, just to give us the Phone Number and from that, the OP can identify the private IP assigned to that phone number by the UPF. this is actually what we need to configure the routing rules on the UPF, isn't it @babunkj. It works for a Smartphone. I think that a device attached to a Fix Network router is out of scope. A device attached to a 5G Router is tricky. Anyway we could clarify that this API works for Smartphones moving around. It is better to have a simple and clear solution for the most common use case than to support any kind of device proving a solution that is difficult to understand for the Developer. What do you think?

**Reply by FabrizioMoggio** [reply]
Clearly having just phone number as an input forces the Service Provider to gather it and it is not always a request the Service Provider wants to do to the Customer.

**Reply by babunkj** [reply]
Giving some views on the above comments of @FabrizioMoggio and @gunjald. 
With some basic understanding of NAT/NAPT as relevant to our use cases, can we state these:
•	The device has an IP address assigned by the network, which is the private IP address. 
•	NAT/NAPT will translate device Private IP Addr+ Port to External IP Addr + Port in the uplink and vice versa for downlink.
•	External IP Addr + Port can always be used to uniquely identify the private IP Addr of the device.  
•	External IP Addr + Port can be mapped to private IP Addr + port too which can help in identifying the specific traffic from the device.  
•	The Application server gets the External IP Addr + Port in the IP packets from the device (via NAT) which it can always use to refer to the device. Even if they change with time, connectivity etc, the application does not need to track them and can always use the External IP Addr + Port as received currently. 

As suggested by @FabrizioMoggio , I also think that for basic and simple cases using Phone number may be easier. A phone number is also the most often used identifier of a device. OP will be able to identify the private IP based on the phone number. 
This is assuming that the Application server is aware of the phone number. Perhaps, there could be situations where the application server is not aware of the phone number and knows only the IP address. For this, an additional option of giving External IP Addr+Port can be provided to cover such cases.

**Comment by FabrizioMoggio** [comment]
Additional information: GSMA OPG.03 defines Public_IP and Source Port to identify the UE. So we can at least have the following conclusion:

**Conclusion:** We need to introduce Source IP port in the TI API

How to, this is still under discussion but I suggest to be aligned with: https://github.com/camaraproject/EdgeCloud/discussions/247

**Reply by gunjald** [reply]
In my view i have a suggestion given in #123 aligned to the QoD API to identify the device uniquely. If we agree with the existing solutions from other APIs and can follow it here we can keep the device identity uniqueness separate from the other behaviors and deal with them separately.

So  device identity issue can be resolved as part of the #123 while the current  issue just focuses on the traffic filters enhancements to cover more granular traffic identification requirements. These scenarios talks about devices with more than one traffic flows but still need to provide different treatment to them based on the filter criteria defined by the application developer in API request.

**Reply by FabrizioMoggio** [reply]
@gunjald I agree with you, isn't it aligned with my previous comment? We can close this discussion with the above conclusion: "We need to introduce Source IP port in the TI API". How to implement it will be defined in #247 that is meant to summarize #123 and other discussions on the same topic.

**Reply by gunjald** [reply]
Yes Fabrizio, it is aligned with your previous comment. May be the terminology we can keep same as used in QoD for "Device" object and which exist even for TI and can be modified with introduction of publicPort. as in QoD "ipv4Address": { "publicAddress": "203.0.113.0", "publicPort": 59765.}

While you can enhance the traffic filters with your proposal on "destintionTrafficFilters and "sourceTrafficFilters:.

Does it look aligns with your understanding?

**Reply by FabrizioMoggio** [reply]
Totally, I have already implemented this in 0.9.5 (not yet published), I still have one doubt before making the PR. I would like to better define the two Ports we have for the UE. The first port is the one in the Device parameter, publicPort. The second one is the port we have in sourceTrafficFilters.

Currently I have this text:

_**sourceTrafficFilters:**  The traffic can be from a specific application port in the Device._

And in Device explanation text, I have: 

_The Device can be identified by the phone number (phoneNumber), an external identifier (networkAccessIdentifier) or by its IP Address (Ipv4Address, Ipv6Address) also specifying a Port._


So I named the first one "application port" and I can't find the right name for the second one (the one in the Device), Maybe "nat port"?

**Reply by gunjald** [reply]
I think "So I named the first one "application port" and I can't find the right name for the second one (the one in the Device), Maybe "nat port"?" we can just use the publicPort same as the QoD as the device can have multiple traffic stream and there can be multiple NATed ports too. But the public Ip and public Port for different stream will still resolve to same device.

As these ports are still under the "Device" object so implicitly they refer to device and we may not need to use terms like application port or nat port in my opinion.

**Reply by FabrizioMoggio** [reply]
maybe I'm lost...let me correct I'm lost for sure :-).

a device can have a Private IP, a Public IP and a Public Port (Port 1)

from an application point of view, it can originate traffic from a different Port (Port 2).

Is this right?

**Reply by gunjald** [reply]
May be it is about too many terms creating confusion. Let me try to put my understanding here.

Device may have an internal IP within internal network dedicated to it. While reaching out to external applications over internet it originates IP packets with source IP as private IP and the internal source port towards the external Public IP and Port of  EAS.

This packet  is translated by the CGNAT which assign a Public IP and Public Port  managed by CGNAT for the Device Private IP and Private Port i.e. source IP and source Port in this conversation. I.e. Device Private IP and Private Port => Public IP, Public Port.(CGNAT).

Please note that the above explanation is totally for device IP and port and nothing has been said about EAS which is assumed to be addressed in the destination IP and Destination Port in the IP packet.

Now based on the different stream of traffic towards EAS the device may have source private IP and private port going via CGNAT and for each unique private, private port for the device CGNAT will allocate unique Public IP, Public Port .

In that case using Public IP, Public Port of the device the private IP and Private Port mapping can be retrieved at CGNAT.

Also, the EAS will see the Device Public Ip and Public Port in the IP packets it receives and may not have visibility of the Private IP and Private port within the internal mobile network. Hence we are proposing to use the Device Public IP and Public Port (could be one or more at same time for different traffic flows) to identify the device.

I hope it aligns with your view.

**Comment by FabrizioMoggio** [comment]
According to the explanation ( thank you, I had it wrong before) I don't think we need sourceTrafficFilter. this is way I wasn't able to differentiate them, they are the same. Am I right?

@babunkj what do you think?

**Reply by gunjald** [reply]
I think we still need sourceTrafficFilter. The reason being that the device may have more than one PublicIP and Port mapped to it depending upon the traffic it is generating towards external application e.g. low resolution stream, high resolution stream.

So that differentiation can be brought in by the sourceTrafficFilter as application developer can just use any of the Public IP and Public Port assigned for the device traffic stream to identify the device. and that combination may not always map to sourceTrafficFilter which is actually needed for traffic influence.

**Reply by gunjald** [reply]
Also, I think sourceTrafficFilter in future could be enhanced to provide more options to identify traffic flows e.g. based on domain names, HTTP keywords as I have seen in some of the 3GPP parameters in ASSessionWithQoS NEF API. But that all can be in future I suppose.

**Reply by FabrizioMoggio** [reply]
ok, so currently I have this text:

_sourceTrafficFilters: The traffic can be from a specific application port in the Device._


_The Device can be identified by the phone number (phoneNumber), an external identifier (networkAccessIdentifier) or by its IP Address (Ipv4Address, Ipv6Address) also specifying a Port._

Is this ok or do you suggest a better wording, naming?

**Reply by babunkj** [reply]
Agree, in Traffic Influence, we need the device identifier to identify the device and an application flow identifier to identify the specific flow(s) to be influenced. This is similar to the QoD API. I think we can keep these two sets of attributes that an application user can easily identify with.

Device Identifier attribute:
Among the multiple ways in which the API Invoker has to identify the device, if it wants to use IP Address, the invoker has to provide the publicAddress and publicPort of Device Ipv4Address. They are as already mentioned by @FabrizioMoggio and @gunjald.

Application flow attribute: 
We will need the source filter which can give the information regarding which application flow is to be influenced. The basic information required can be:
-	Application server address and port 
-	Device port(s) 
The device ports mentioned here are the private ports (not natted) that the device is using for the flows to be influenced. This port need not be same as the port that the application server has received currently before influencing. For example, the application server may receive a Low resolution flow (Port A) but wants to influence a high resolution flow (Port B). In this case, the invoker does not have any info about natted port for the High resolution flow.
In QoD API, these ports are referred as devicePorts. I think the explanation for the devicePorts attribute in QoD API may be applied for TI too. I have just replaced QoS phrases with Traffic influence ones – “The ports used locally by the device for flows to which the requested traffic influence should apply. If omitted, then the traffic influence will apply to all flows between the device and the specified application server address and ports.”.

**Reply by FabrizioMoggio** [reply]
I agree on aligning with QoD, thank you @babunkj

**Reply by FabrizioMoggio** [reply]
I'm wondering, @babunkj, you wrote Port(s). In QoD they actually define multiple ports. Currently for the destination Port we have just one port. Do we need to have multiple ports as source or as destination? The API Consumer can anyway invoke the API for each port. But maybe it is better to support multiple ports. So we could align with QoD using "devicePorts". If yes, for both "source" and "destination"?

**Reply by babunkj** [reply]
Yes @FabrizioMoggio I feel it is better to have a possibility of a range of ports for source and destination ports. Multiple ports will help identify a category of traffic to be influenced.

**Comment by FabrizioMoggio** [comment]
Implemented with: https://github.com/camaraproject/EdgeCloud/pull/249

**Comment by FabrizioMoggio** [comment]
Merged rel 0.9.5

**Comment by FabrizioMoggio** [comment]
Actually not all is implemented in Rel 0.9.5: https://github.com/camaraproject/EdgeCloud/discussions/230#discussioncomment-9646458

**Reply by FabrizioMoggio** [reply]
The Device parameter is as defined in: https://github.com/camaraproject/Commonalities/blob/main/artifacts/CAMARA_common.yaml 
 is it better to be coherent with it or to modify it to have device port as "list"? I suggest to stick with the one defined in Commonalities. The API Consumer can anyway invoke the TI API more times to cover all the ports. It is better to have the API simple and coherent with Commonalities rather than covering all the possible use cases.

**Reply by javierlozallu** [reply]
As was commented on today's meeting, Device Object will be updated in commonalities v0.4.0, so I agree to be coherent with it

**Comment by FabrizioMoggio** [comment]
Proposed solution according to: https://github.com/camaraproject/EdgeCloud/pull/249#issuecomment-2178016648

Proposal:

we use Device as defined in Commonalities
we add sourceTrafficFilters to provide the option to add more PublicPorts
we keep destinationTrafficFilters
we add a good explanation in the Documentation :-)

**Reply by gunjald** [reply]
Good from my side.

**Reply by babunkj** [reply]
Agree

**Reply by FabrizioMoggio** [reply]
...ehm...sorry but I need a further question :-)

destinationTrafficFilters: currently it is just one Port-Protocol. What about having also it as a list?

**Reply by gunjald** [reply]
One point i think we also need to think is about the handling of multiple ports in any filter that the API consumer should be able to retrieve the status of influence each port. As the API will be aggregating multiple ports it will required to be known what happens to each port. Quite possible that each port may not get the treatment it was asked for so aggregating status of individual port may not be possible.

**Reply by FabrizioMoggio** [reply]
So your suggestion is to keep destinationTrafficFilters as just one port/protocol and if the Service exposes different ports/protocols the API Consumer invokes the TI API for each of them so their LCM can be done independently. Is my understanding correct?

**Reply by gunjald** [reply]
I mean to say we can keep the multiple ports as were discussing but then provide the capability in GET method to query status of influence based on individual port and source/destination combination. Any thoughts?

**Reply by FabrizioMoggio** [reply]
So, this is my understanding:

- we ca have multiple ports for both sourceTrafficFilters and destinationTrafficFilters.
- we need to provide back the status for each flow

one comment, if we have multiple ports for both source and destination, we will influence any kind possible flow, mixing all the sources and all the destinations. Example:

source: S_Port_1 and S_Port_2
destination: D_Port_1 and D_Port_2

The configuration on the UPF will be:

- from S_Port_1 to D_Port_1 
- from S_Port_1 to D_Port_2
- from S_Port_2 to D_Port_1 
- from S_Port_2 to D_Port_2

if the API Consumer wants something different, it needs to make more TI API calls. For example if the intent is

- from S_Port_1 to D_Port_1 
- from S_Port_2 to D_Port_2

the API Consumer needs two TI API calls:

- TI-API-Call-1 with S_Port_1and D_Port_1
- TI-API-Call-2 with S_Port_2and D_Port_2

If we keep destinationTrafficFilter as just one Port, we force the API Consumer to always make more API calls but, maybe, it is more straightforward and easy to grasp.

What do you think about?

**Reply by babunkj** [reply]
I had a slightly different view on the traffic filters.
- My understanding is that the sourceTrafficFilters  and destinationTrafficFilter together identify the IP traffic to be influenced. 
For example, when each of the filters have one port, it means the IP traffic with source as ‘device-IPAddress + sourcePort’ and destination as ‘applicationServer + destinationPort’ will be diverted to the cloudlet corresponding to the best Region-Zone applicable (Specific Region-Zone if requested in the API or the best Region-Zone as identified by OP).
When the sourceTrafficFilters  and destinationTrafficFilter have a range of ports, any of the IP traffic flows with the source values within  the ‘device-IPAddress+range of sourcePorts’ and destination values within the ‘applicationServer + range of destinationPorts’ will be diverted to the cloudlet corresponding to the best Region-Zone applicable (Specific Region-Zone if requested in the API or the best Region-Zone as identified by OP).
-	With the above logic,  the influencing is done in the same way with all the traffic corresponding to each port in the port range. Ie all these traffic are detected by UPF (using the source-destination IP Address-port values) and routed through the N6 interface to the Cloudlet corresponding to the Region-Zone.
-	If the Application wants to divert two different traffic flows to two different cloudlets, it has to invoke two different Traffic Influence requests.

**Reply by gunjald** [reply]
Yes in my opinion also having one call for a flow is simple and more manageable. With multiple ports one issue is that the API shall also provide a way to invoker to specify which source port and which destination port represent one logical traffic flow from application point of view. It cannot be true that all the combinations of source and destination ports will be valid as many of them may not be representing the valid traffic flow. Which means then API should also enable identification of valid traffic flows that are logically valid among the given source and destination ports. Then the next challenge is to aggregate the traffic influence status for each such pair from given set of logically valid traffic flows.

So in my opinion simplistic approach may be good for now and in future we may evolve the other possible models too.

**Reply by FabrizioMoggio** [reply]
@babunkj and @gunjald : I'm not sure your comments leads to the same solution :-)

> I had a slightly different view on the traffic filters.
> 
> * My understanding is that the sourceTrafficFilters  and destinationTrafficFilter together identify the IP traffic to be influenced.
>   For example, when each of the filters have one port, it means the IP traffic with source as ‘device-IPAddress + sourcePort’ and destination as ‘applicationServer + destinationPort’ will be diverted to the cloudlet corresponding to the best Region-Zone applicable (Specific Region-Zone if requested in the API or the best Region-Zone as identified by OP).
>   When the sourceTrafficFilters  and destinationTrafficFilter have a range of ports, any of the IP traffic flows with the source values within  the ‘device-IPAddress+range of sourcePorts’ and destination values within the ‘applicationServer + range of destinationPorts’ will be diverted to the cloudlet corresponding to the best Region-Zone applicable (Specific Region-Zone if requested in the API or the best Region-Zone as identified by OP).
> * With the above logic,  the influencing is done in the same way with all the traffic corresponding to each port in the port range. Ie all these traffic are detected by UPF (using the source-destination IP Address-port values) and routed through the N6 interface to the Cloudlet corresponding to the Region-Zone.
> * If the Application wants to divert two different traffic flows to two different cloudlets, it has to invoke two different Traffic Influence requests.

@babunkj your comments starts with "I had a slightly different view on the traffic filters". Actually,  I can't see the difference with respect to my example. I mean the example where we have a list of ports for sourceTrafficFilters  and destinationTrafficFilter. Please can you help me to detect the difference?

**Reply by FabrizioMoggio** [reply]
> Yes in my opinion also having one call for a flow is simple and more manageable. With multiple ports one issue is that the API shall also provide a way to invoker to specify which source port and which destination port represent one logical traffic flow from application point of view. It cannot be true that all the combinations of source and destination ports will be valid as many of them may not be representing the valid traffic flow. Which means then API should also enable identification of valid traffic flows that are logically valid among the given source and destination ports. Then the next challenge is to aggregate the traffic influence status for each such pair from given set of logically valid traffic flows.
> 
> So in my opinion simplistic approach may be good for now and in future we may evolve the other possible models too.

So you are suggesting, for sake of simplicity, in this release, to have just one port for destinationTrafficFilter. This because maybe all the permutations are not valid. Is this correct? 

We can still have sourceTrafficFilters as list of ports. In this case the API Consumer can ask, with no ambiguity, for flows from those source ports to the one destinatin port specified in destinationTrafficFilter.

**Reply by gunjald** [reply]
For "So you are suggesting, for sake of simplicity, in this release, to have just one port for destinationTrafficFilter. This because maybe all the permutations are not valid. Is this correct?", my answer is "YES"

For "We can still have sourceTrafficFilters as list of ports. In this case the API Consumer can ask, with no ambiguity, for flows from those source ports to the one destinatin port specified in destinationTrafficFilter.", my answer is "Not sure if there would be cases where one or more source ports are having a relation with single destination port." In that case even single source port may serve many use cases just that app developers will be need to target one flow at a time. Basically we need assess if such scenario exists for one or more source ports and a single destination port. If not then we can just one port each in source and destination filters in my view.

**Reply by FabrizioMoggio** [reply]
> For "So you are suggesting, for sake of simplicity, in this release, to have just one port for destinationTrafficFilter. This because maybe all the permutations are not valid. Is this correct?", my answer is "YES"
> 
> For "We can still have sourceTrafficFilters as list of ports. In this case the API Consumer can ask, with no ambiguity, for flows from those source ports to the one destinatin port specified in destinationTrafficFilter.", my answer is "Not sure if there would be cases where one or more source ports are having a relation with single destination port." In that case even single source port may serve many use cases just that app developers will be need to target one flow at a time. Basically we need assess if such scenario exists for one or more source ports and a single destination port. If not then we can just one port each in source and destination filters in my view.

I tend to agree with you, I have the same feeling. Just one source port is enough for must use cases.

**Reply by babunkj** [reply]
> @babunkj your comments starts with "I had a slightly different view on the traffic filters". Actually, I can't see the difference with respect to my example. I mean the example where we have a list of ports for sourceTrafficFilters and destinationTrafficFilter. Please can you help me to detect the difference?

@FabrizioMoggio  This is the view I was having of the example you had taken. Perhaps it is similar to the way you had thought too. 

You had mentioned the below scenario:
source: S_Port_1 and S_Port_2
destination: D_Port_1 and D_Port_2
The configuration on the UPF will be:
•	from S_Port_1 to D_Port_1
•	from S_Port_1 to D_Port_2
•	from S_Port_2 to D_Port_1
•	from S_Port_2 to D_Port_2

I was interpreting your comment as saying that UPF will route from S_Port_1 to D_Port_1 OR from S_Port_1 to D_Port_2. 

But as I understand, it is not that UPF will route S_Port_1 to D_Port_1 OR S_Port_1 to D_Port_2 etc.  UPF will apply same N6 traffic routing rules on all the traffic flows that have source as S_Port_1 or S_Port_2 and destination as D_Port_1 or D_Port_2. UPF will be configured to look for any traffic flow which has S_Port_1 or S_Port_2 as source AND D_Port_1 or D_Port_2 as destination. Once any of the traffic flows which matches these reaches 5G Core (in the UL direction), UPF will route the traffic to the N6 configured for it in the UPF rules – which is towards the nearest Edge Cloud. 

Thus, the traffic influence applied is exactly same for all the 4 traffic flows.

**Reply by FabrizioMoggio** [reply]
exactly we have the same understanding

**Comment by FabrizioMoggio** [comment]
please note that if we go for just one source port, sourceTrafficFilters, if used, must be equal to the port defined in Device, if used. What happens if are used two different  ports? We can document that if they are both valorised, if not equal, the one defined in sourceTrafficFilters, will "win".

**Reply by gunjald** [reply]
I think a device application can have more than one port that it uses and both identifies the same device along with the device IP address. So practically the device out of multiple ports may just want to influence a specific port and that may be different than the port which is used for the device identification. 

So in my opinion we should not put this restriction of having same port in Device attribute and in the sourceTrafficFilters. Any thoughts?

**Reply by FabrizioMoggio** [reply]
you are totally right, I'm loosing track of the many ports :-)

so:

- device-port and sourceTrafficFilters ports can be different

question:

- if they are both specified (and different), is only sourceTrafficFilters used? (or do we use both?)
-  if only device-port is specified, is it used? (I think so)


anyway we still need to decide if sourceTrafficFilters is a list or not.

**Reply by javierlozallu** [reply]
I agree with Deepak that ports could be different or the same. I understand that the sourceIp:sourcePort Device is just to identify the device and the source/destination traffic filters to identify which Traffic Flow will be affected by TI API.

So although the ports are the same, they have different purposes. 

If only device-port is used, it will be assumed that this is the port of the traffic flow that will be influenced by the TI API.

**Reply by babunkj** [reply]
Agree with above points. 
-	Application can be related to more than one source port
-	device-port and sourceTrafficFilters ports can be different. The Application may want to influence a specific port which may be different than the port which is used for the device identification. (ex. High resolution video feed to be switched based on the analysis of Low resolution video feed)

We do have use case for single port in sourceTrafficFilters  and a destination port in destinationTrafficFilter. The camera use case mentioned here can be actually extended to say ‘switch all the high-resolution video feeds (multiple source ports) to the Edge video server based on video analysis of a low-resolution video feed’ – which would mean multiple source ports in sourceTrafficFilters. 

I am also in agreement going for single source and destination port initially and then extending in future to multiple ports based on more use cases being brought up in the group.

**Comment by FabrizioMoggio** [comment]
According to the above discussion:

Conclusion:

- we use Device as defined in Commonalities
- we add sourceTrafficFilters to provide the option to add one source PublicPort
- we keep destinationTrafficFilters, with one destination Port
- device-port can be different from sourceTrafficFilters-port
- if sourceTrafficFilters-port is valorised, it is used for to influence the traffic. device-port is not used to influence the traffic
- if sourceTrafficFilters-port is not valorised and device-port is valorised, device-port  is used to influence the traffic
- we add a good explanation in the Documentation :-)

**Reply by babunkj** [reply]
Agree

**Comment by FabrizioMoggio** [comment]
implemented in: https://github.com/camaraproject/EdgeCloud/pull/278
